Kubernetes network policy controller

awh opened the pull request:
For initial code review; documentation + integration tests to follow.

awh committed:
Kubernetes network policy controller

awh committed:
Added diagnostic logging

brb commented on cmd/weave-npc/main.go:
```diff
@@ -0,0 +1,160 @@
+				handleError(npc.UpdateNetworkPolicy(old.(*extensions.NetworkPolicy), new.(*extensions.NetworkPolicy)))
+			}})
+
+	go nsController.Run(wait.NeverStop)
```
Does the k8s event handler deliver the next event after the previous event has been sent to a channel? Or does it wait until a callback associated with an event has finished executing?

If the former is true, then it might be not safe to handle events in separate goroutines due to possible races.

awh replied:
There's a mutex in the `NetworkPolicyController` that serialises event processing.

brb commented on cmd/weave-npc/main.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+func main() {
+
```
Nit picking - empty space.

awh replied:
:thumbsup:

brb commented on cmd/weave-npc/main.go:
```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	weavenpc "github.com/weaveworks/weave-npc/pkg/controller"
```
Nit picking - you might want to group deps into the following:

```
stdlib deps

3rd party deps

local deps
```

It applies to other modules as well.

awh replied:
:+1:

brb commented on pkg/util/ipset/ipset.go:
```diff
@@ -0,0 +1,72 @@
+type ipset struct {
+}
+
+func New() Interface {
```
I'd return `*ipset`.

awh replied:
This package is modelled on `k8s.io/kubernetes/pkg/util/iptables` so I've followed the convention there.

brb commented on pkg/util/ipset/ipset.go:
```diff
@@ -0,0 +1,72 @@
+package ipset
+
+import (
+	"github.com/pkg/errors"
```
:+1:

brb commented on cmd/weave-npc/main.go:
```diff
@@ -0,0 +1,160 @@
+	}
+
+	if _, err := ipt.EnsureRule(iptables.Append, iptables.TableFilter, weavenpc.MainChain,
+		"-m", "state", "--state", "NEW", "-j", "WEAVE-NPC-DEFAULT"); err != nil {
```
Nit picking - maybe use `{Main,Default,Ingress}Chain` constants instead of raw strings?

awh replied:
:thumbsup:

brb commented on pkg/controller/selector.go:
```diff
@@ -0,0 +1,124 @@
+type selectorSet struct {
+	ips           ipset.Interface
+	onNewSelector selectorFn
+	users         map[string]map[types.UID]struct{} // list of users per selector
```
It's not clear (at least to me) what is a "user" in this context?

awh replied:
They're the UIDs of k8s objects (usually network policies) that depend on this selector. Open to a better name!

brb commented on docs/DESIGN.markdown:
```diff
@@ -0,0 +1,90 @@
+
+## Dynamically maintained `WEAVE-NPC-DEFAULT` chain
+
+The policy controller maintains a rule in this chain for every
```
Feel free to ignore (probably it is already set in stone by the k8s network policy proposal), but wouldn't it make sense to allow a custom rule to override a default policy?

E.g. in iptables a default policy always applies at the end of a chain, so a user is able to add a custom rule to prevent a default policy from taking an action.

awh replied:
As you suspect, the k8s network policy defines the default as `DefaultAllow`. I think you're right though, people may want to fiddle around with things like this - let's see what people ask for.

brb commented on docs/DESIGN.markdown:
```diff
@@ -0,0 +1,90 @@
+destroyed automatically in response to Pod, Namespace and
+NetworkPolicy object updates from the k8s API server:
+
+* A `hash:ip` set per namespace, containing the IP addresses of all
```
IMHO, for sake of readability, you could add here references to ipsets used in the code.

awh replied:
I am not sure what kind of reference you mean here.

brb commented on pkg/controller/controller.go:
```diff
@@ -0,0 +1,161 @@
+
+func (npc *controller) onNewNsSelector(selector *selector) error {
+	for _, ns := range npc.nss {
+		if ns.namespace != nil {
```
When can it be `nil`?

awh replied:
There are no ordering guarantees regarding when we hear about new Namespace objects vs say Pods that inhabit that namespace - we could have created the `ns` struct because of a pod creation, and then only fill in the k8s Namespace object later on.

awh committed:
Build git version into executable

awh committed:
Update documentation

awh committed:
Tolerate NoSchedule taint

brb commented on docs/DESIGN.markdown:
```diff
@@ -0,0 +1,100 @@
+
+To direct traffic into the policy engine:
+
+iptables -A FORWARD -o weave -m physdev ! --physdev-out vethwe-bridge -j WEAVE-NPC
```
Where does it get installed?

awh replied:
These rules have to happen outside of weave-npc to avoid a race at startup and a potential fail-open condition if the policy engine fails to start. They'll be added by the weave script when it is run from weave-kube.

brb reviewed (commented):
Initial review.

awh committed:
Remove DROP from static chain

We expect this to be configured outside of weave NPC (e.g. by weave-kube
or the node administrator)

awh committed:
Group imports
